include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -185,29 +185,82 @@ struct __init_processing
 //------------------------------------------------------------------------
 
 template <typename _ExecutionPolicy, typename _Operation1, typename _Operation2>
-struct transform_init
+struct transform_init_seq
 {
     _Operation1 __binary_op;
     _Operation2 __unary_op;
 
-    template <typename _NDItemId, typename _Size, typename _AccLocal, typename... _Acc>
+    template <typename _Size, typename _AccLocal, typename... _Acc>
     void
-    operator()(const _NDItemId __item, _Size __n, ::std::size_t __iters_per_work_item, ::std::size_t __global_id,
-               ::std::size_t __global_offset, _AccLocal& __local_mem, const _Acc&... __acc) const
+    operator()(const _Size __n, _AccLocal& __local_mem, const _Acc&... __acc) const
+    {
+        __local_mem = __unary_op(0, __acc...);
+        // Add neighbour to the current __local_mem
+        for (_Size __i = 1; __i < __n; ++__i)
+            __local_mem = __binary_op(__local_mem, __unary_op(__i, __acc...));
+    }
+};
+
+template <typename _ExecutionPolicy, ::std::size_t __iters_per_work_item, typename _Operation1, typename _Operation2>
+struct transform_init_known
+{
+    _Operation1 __binary_op;
+    _Operation2 __unary_op;
+
+    template <typename _Size, typename _AccLocal, typename... _Acc>
+    void
+    operator()(const ::std::uint16_t __local_id, const _Size __n,
+               const ::std::size_t /* unused __iters_per_work_item */, const ::std::size_t __global_id,
+               const ::std::size_t __global_offset, _AccLocal& __local_mem, const _Acc&... __acc) const
+    {
+        const ::std::size_t __adjusted_global_id = __global_offset + __iters_per_work_item * __global_id;
+        const ::std::size_t __items_to_process = __n - (__iters_per_work_item * __global_id);
+        // Add neighbour to the current __local_mem
+        if (__items_to_process >= __iters_per_work_item)
+        {
+            // Keep these statements in the same scope to allow for better memory alignment
+            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
+#pragma unroll
+            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
+                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __i, __acc...));
+            __local_mem[__local_id] = __res;
+        }
+        else if (__items_to_process > 0)
+        {
+            // Keep these statements in the same scope to allow for better memory alignment
```

Comment:
Lets comment(explain) why we have absolutely identical lines as above.

Reply:
I have unified the sequential and `transform_reduce_known` structures and introduced a function `apply_unary_binary` to reduce the duplicated code. There is also a comment on vectorization.